### PR TITLE
[Feature] WaterButler move n copy

### PR DESCRIPTION
**Pull request by chrisseto**

### CODE REVIEW

File: website/addons/osfstorage/decorators.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            raise HTTPError(httplib.CONFLICT)
+    return wrapped
+
+
+def autoload_filenode(must_be=None, default_root=False):
+    def _autoload_filenode(func):
+        @handle_odm_errors
+        @functools.wraps(func)
+        def wrapped(*args, **kwargs):
+            node_addon = kwargs['node_addon']
```
Comment (sloria) [inline website/addons/osfstorage/decorators.py]: Can you add documentation to indicate that this decorator must be used together with a decorator that injects `node_addon`?

File: website/addons/osfstorage/decorators.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+            if 'fid' not in kwargs and default_root:
+                file_node = node_addon.root_node
+            else:
+                file_node = model.OsfStorageFileNode.get(kwargs.get('fid'), node_addon)
+
+            if must_be and file_node.kind != must_be:
+                raise HTTPError(httplib.BAD_REQUEST)
```
Comment (sloria) [inline website/addons/osfstorage/decorators.py]: Pass a useful error `message_long` in the `data` argument.

File: website/addons/osfstorage/decorators.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def waterbutler_opt_hook(func):
+
+    @must_be_signed
+    @handle_odm_errors
+    @must_not_be_registration
+    @must_have_addon('osfstorage', 'node')
+    @functools.wraps(func)
+    def wrapped(payload, *args, **kwargs):
```
Comment (sloria) [inline website/addons/osfstorage/decorators.py]: Where does `payload` come from? `must_be_signed`? Add a comment about where this kwarg comes from.

File: website/addons/base/views.py [hunk 4]
```diff
@@ -178,27 +181,58 @@ def get_auth(**kwargs):
...
+
     auth = Auth(user=user)
-    node_addon.create_waterbutler_log(auth, osf_action, metadata)
+    node = kwargs['node'] or kwargs['project']
+
+    if action in (NodeLog.FILE_MOVED, NodeLog.FILE_COPIED):
+        for bundle in ('source', 'destination'):
+            for key in ('provider', 'materialized', 'name'):
+                if key not in payload[bundle]:
+                    raise HTTPError(httplib.BAD_REQUEST)
```
Comment (sloria) [inline website/addons/base/views.py]: Pass a useful error message_long here.

File: website/addons/base/views.py [hunk 5]
```diff
@@ -178,27 +181,58 @@ def get_auth(**kwargs):
...
+
+        source = node.get_addon(payload['source']['provider'])
```
Comment (sloria) [inline website/addons/base/views.py]: Can `source` ever be `None`?

**Comment by sloria** [comment]
@chrisseto Which waterbutler branch does this depend on?

**Comment by chrisseto** [comment]
@sloria https://github.com/CenterForOpenScience/waterbutler/pull/31

File: website/addons/base/views.py (hunk 5 above)
Comment (chrisseto) [inline website/addons/base/views.py]: Not unless something terrible is happening.

File: website/addons/osfstorage/views.py [hunk 6]
```diff
@@ -38,250 +54,197 @@ def make_error(code, message_short=None, message_long=None):
...
-                    name=folder.name,
-                    path=folder.materialized_path(),
-                )
-            })
-        folder.undelete(Auth(user), recurse=False)
-    folder.log(Auth(user), NodeLog.FOLDER_CREATED)
-
-    return folder.serialized(), httplib.CREATED
+# lol
+for key, item in locals().items():
```
Comment (sloria) [inline website/addons/osfstorage/views.py]: This is unnecessary. Why not just decorate the functions explicitly?

File: website/static/js/fangorn.js [hunk 7]
```diff
@@ -236,6 +250,105 @@ function _fangornToggleCheck(item) {
...
+        from = tb.createItem($.extend(true, {}, from.data), to.id);
+        from.data.status = 'copy';
+    }
+
+    tb.redraw();
+
+    $.ajax({
+        type: 'POST',
+        url: isMove ? waterbutler.moveUrl() : waterbutler.copyUrl(),
+        headers: {
```
Comment (sloria) [inline website/static/js/fangorn.js]: You an use `osfHelpers.postJSON` here.

File: website/static/js/fangorn.js [hunk 8]
```diff
@@ -236,6 +250,105 @@ function _fangornToggleCheck(item) {
...
+                child = tb.buildTree(item, from);
+                inheritFromParent(child, from);
+                from.add(child);
+            });
+            from.open = true;
+            from.load = true;
+        }
+
+        tb.redraw();
+    }).fail(function() {
```
Comment (sloria) [inline website/static/js/fangorn.js]: Log this error to Sentry.

File: website/static/js/fangorn.js [hunk 9]
```diff
@@ -236,6 +250,105 @@ function _fangornToggleCheck(item) {
...
+        if (isMove) {
+            from.move(ogParent);
+            from.data.status = undefined;
+        } else {
+            from.removeSelf();
+        }
+
+        $osf.growl('Error', (isMove ? 'Move' : 'Copy') + ' failed.');
```
Comment (sloria) [inline website/static/js/fangorn.js]: This error message could be more descriptive. At the very least, include the support email, e.g. "If the problem persists, please contact support@osf.io".

**Comment by sloria** [comment]
Merged into `release/0.37.0`

**Pull request closed without merging**